File: Snooper/src/androidTest/java/com/prateekj/snooper/rules/TestDbRule.java
package com.prateekj.snooper.rules;


import android.content.Context;
import android.content.ContextWrapper;
import androidx.annotation.NonNull;
import androidx.annotation.RawRes;
import androidx.test.InstrumentationRegistry;

import com.prateekj.snooper.utils.Logger;

import org.junit.rules.TestRule;
import org.junit.runner.Description;
import org.junit.runners.model.Statement;

import java.io.File;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;

public class TestDbRule implements TestRule {
  private int dbRawResourceId;
  private String dbName;

  public TestDbRule(@RawRes int dbRawResourceId, String dbName) {
    this.dbRawResourceId = dbRawResourceId;
    this.dbName = dbName;
  }

  @Override
  public Statement apply(final Statement base, Description description) {
    return new Statement() {
      @Override
      public void evaluate() throws Throwable {
        copyDataBase(TestDbRule.this.dbName);
        base.evaluate();
        InstrumentationRegistry.getTargetContext().deleteDatabase(dbName);
      }
    };
  }

  private void copyDataBase(String finalDbName) {
    Context applicationContext = InstrumentationRegistry.getTargetContext();
    ContextWrapper cw = new ContextWrapper(applicationContext);
    File dbDir = new File(getDBDirectory());
    if(!dbDir.exists()) {
      dbDir.mkdir();
    }
    Logger.d("Database", "New database is being copied to device!");
    byte[] buffer = new byte[1024];
    int length;
    try {
      InputStream myInput = applicationContext.getResources().openRawResource(this.dbRawResourceId);
      File file = new File(dbDir , finalDbName);
      OutputStream myOutput = new FileOutputStream(file);
      while ((length = myInput.read(buffer)) > 0) {
        myOutput.write(buffer, 0, length);
      }
      myOutput.close();
      myOutput.flush();
      myInput.close();
      Logger.d("Database", "New database has been copied to device!");
    } catch (IOException e) {
      Logger.e("Database", e.getMessage(), e);
    }
  }

  @NonNull
  public String getDBDirectory() {
    ContextWrapper cw = new ContextWrapper(InstrumentationRegistry.getTargetContext());
    String destPath = cw.getFilesDir().getPath();
    return destPath.substring(0, destPath.lastIndexOf("/")) + "/databases";
  }
}

File: Snooper/src/androidTest/java/com/prateekj/snooper/rules/TestDbRule.kt
package com.prateekj.snooper.rules


import android.content.ContextWrapper
import androidx.annotation.RawRes
import androidx.test.platform.app.InstrumentationRegistry.getInstrumentation
import com.prateekj.snooper.utils.Logger
import org.junit.rules.TestRule
import org.junit.runner.Description
import org.junit.runners.model.Statement
import java.io.File
import java.io.FileOutputStream
import java.io.IOException

class TestDbRule(@param:RawRes private val dbRawResourceId: Int, private val dbName: String) :
  TestRule {

  val dbDirectory: String
    get() {
      val cw = ContextWrapper(getInstrumentation().targetContext)
      val destPath = cw.filesDir.path
      return destPath.substring(0, destPath.lastIndexOf("/")) + "/databases"
    }

  override fun apply(base: Statement, description: Description): Statement {
    return object : Statement() {
      @Throws(Throwable::class)
      override fun evaluate() {
        copyDataBase(this@TestDbRule.dbName)
        base.evaluate()
        getInstrumentation().targetContext.deleteDatabase(dbName)
      }
    }
  }

  private fun copyDataBase(finalDbName: String) {
    val applicationContext = getInstrumentation().targetContext
    val dbDir = File(dbDirectory)
    if (!dbDir.exists()) {
      dbDir.mkdir()
    }
    Logger.d("Database", "New database is being copied to device!")
    try {
      val myInput = applicationContext.resources.openRawResource(this.dbRawResourceId)
      val file = File(dbDir, finalDbName)
      val myOutput = FileOutputStream(file)

      myInput.use { inputStream ->
        myOutput.use { outputStream ->
          inputStream.copyTo(outputStream)
        }
      }

      Logger.d("Database", "New database has been copied to device!")
    } catch (e: IOException) {
      Logger.e("Database", e.message, e)
    }
  }
}
